pkg/platform/aks/aks.go

```diff
@@ -0,0 +1,354 @@
+// Azure AKS.
+package aks
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"text/template"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/gohcl"
+	"github.com/mitchellh/go-homedir"
+
+	"github.com/kinvolk/lokomotive/pkg/platform"
+	"github.com/kinvolk/lokomotive/pkg/platform/util"
+	"github.com/kinvolk/lokomotive/pkg/terraform"
+)
+
+// workerPool defines "worker_pool" block.
+type workerPool struct {
+	// Label field.
+	Name string `hcl:"name,label"`
+
+	// Block properties.
+	Count  int               `hcl:"count,optional"`
+	VMSize string            `hcl:"vm_size,optional"`
+	Labels map[string]string `hcl:"labels,optional"`
+	Taints []string          `hcl:"taints,optional"`
+}
+
+// config defines "cluster" block for AKS.
+type config struct {
+	AssetDir    string            `hcl:"asset_dir,optional"`
+	ClusterName string            `hcl:"cluster_name,optional"`
+	Tags        map[string]string `hcl:"tags,optional"`
+
+	// Azure specific.
+	TenantID       string `hcl:"tenant_id,optional"`
+	SubscriptionID string `hcl:"subscription_id,optional"`
+	ClientID       string `hcl:"client_id,optional"`
+	ClientSecret   string `hcl:"client_secret,optional"`
+
+	Location string `hcl:"location,optional"`
+
+	// ApplicationName for created service principal.
+	ApplicationName string `hcl:"application_name,optional"`
+
+	ResourceGroupName   string `hcl:"resource_group_name,optional"`
+	ManageResourceGroup bool   `hcl:"manage_resource_group,optional"`
+
+	WorkerPools []workerPool `hcl:"worker_pool,block"`
+}
+
+const (
+	name = "aks"
+
+	// Environment variables use to load sensitive parts of the configuration.
+	clientIDEnv       = "LOKOMOTIVE_AKS_CLIENT_ID"
+	clientSecretEnv   = "LOKOMOTIVE_AKS_CLIENT_SECRET" // #nosec G101
+	subscriptionIDEnv = "LOKOMOTIVE_AKS_SUBSCRIPTION_ID"
+	tenantIDEnv       = "LOKOMOTIVE_AKS_TENANT_ID"
+)
+
+// init registers aks as a platform.
+func init() { //nolint:gochecknoinits
+	c := &config{
+		Location:            "West Europe",
+		ManageResourceGroup: true,
+	}
+
+	platform.Register(name, c)
+}
+
+// LoadConfig loads configuration values into the config struct from given HCL configuration.
+func (c *config) LoadConfig(configBody *hcl.Body, evalContext *hcl.EvalContext) hcl.Diagnostics {
+	if configBody == nil {
+		emptyConfig := hcl.EmptyBody()
+		configBody = &emptyConfig
+	}
+
+	if d := gohcl.DecodeBody(*configBody, evalContext, c); d.HasErrors() {
+		return d
+	}
+
+	return c.checkValidConfig()
+}
+
+// checkValidConfig validates cluster configuration.
+func (c *config) checkValidConfig() hcl.Diagnostics {
+	var d hcl.Diagnostics
+
+	// Collect all validation functions here.
+	f := []func() hcl.Diagnostics{
+		c.checkNotEmptyWorkers,
+		c.checkWorkerPoolNamesUnique,
+		c.checkWorkerPools,
+		c.checkCredentials,
+		c.checkRequiredFields,
+	}
+
+	// Then call all of them and collect the diagnostics.
+	for _, vf := range f {
+		d = append(d, vf()...)
+	}
```

Comment:
why this function and then the range? It seems more straightforward and readable to just use:
```
d  = append(d, c.checkNot...)
d  = append(d, c.checkWorker...)
```

And is even shorter.

Reply:
It feels like less typing when adding new validation rules to me. Not sure which one is better to be honest.

Reply:
I find it overly complicated having a closure with functions and then go over them, instead of running them in one line.

And when adding, is just c&p and modify the function name, just as in the code that is written today. I don't think there is much less typing.

I prefer clear straightforward code even if it a little bit more typing. Specially in this case, that all is used right away. Is not that we need to receive arbitrary function pointers as argument or something and then run them. We know them in advance, statically, and we just need to run them.

Reply:
Sure, will change it then.

Reply:
Thanks! Let me know if it seems wrong after doing it :)

Reply:
Done.